### PR TITLE
Test for function rotateElement. PashaBeshchuk/tetris#9

### PR DESCRIPTION
**Pull request by PashaBeshchuk**

### CODE REVIEW

File: test/test-rotateElement.js [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+var tetris = require('../tetris')
+describe("element rotation around the axis", function () {
+    var elementTetromino = tetris.getCoordinates(`
+		---
+		--X
+		---
+	`)
+
+    it("element move up one from the right", function () {
+        let stepElementTetromino = tetris.rotateElement(elementTetromino)
```
Comment (zbstof) [inline test/test-rotateElement.js]: This function should accept 2 parameters: coorditates of coordinate to rotate and "pivot" around which rotation is performed

Comment (zbstof) [reply test/test-rotateElement.js]: Please apply to all invocations

Comment (PashaBeshchuk) [reply test/test-rotateElement.js]: Done

File: test/test-rotateElement.js [hunk 2]
```diff
@@ -0,0 +1,203 @@
...
+			---
+			---
+        `)
+        assert.deepEqual(expectedResult, stepElementTetromino)
+    })
+
+    it("element move the left", function () {
+        let elementTetromino = tetris.getCoordinates(`
+            ---
+            --X
```
Comment (zbstof) [inline test/test-rotateElement.js]: Incorrect
Exactly the same input as in previous test. Please check all your tests

Comment (PashaBeshchuk) [reply test/test-rotateElement.js]: I thought it should be so.
I will turn the same coordinate.

Comment (zbstof) [reply test/test-rotateElement.js]: No
All tests should have different inputs and different outputs 
Currently, you have same inputs and different output, which is impossible
Please change inputs to all your tests

Comment (PashaBeshchuk) [reply test/test-rotateElement.js]: Done

File: test/test-rotateElement.js [hunk 3]
```diff
@@ -0,0 +1,203 @@
...
+            ---
+        `)
+
+        let fulcrum = tetris.getCoordinates(`
+            ---
+            -X-
+            ---
+        `)
+        let stepElementTetromino = tetris.rotateElement(fulcrum, elementTetromino)
```
Comment (zbstof) [inline test/test-rotateElement.js]: Please make rotateElement inputs be objects of coordinates, instead ifarrays of such objects
You can take zeroes element of the result of getCoordinates

Comment (zbstof) [reply test/test-rotateElement.js]: Apply to all places

File: test/test-rotateElement.js [hunk 4]
```diff
@@ -0,0 +1,203 @@
...
+		let expectedResult = tetris.getCoordinates(`
+            -----
+            -----
+            ----X
+            -----
+            -----
+        `)
+        assert.deepEqual(expectedResult, stepElementTetromino)
+    })
+})
```
Comment (zbstof) [inline test/test-rotateElement.js]: Add more tests - input should have different x AND y from pivot

Comment (zbstof) [reply test/test-rotateElement.js]: 2 cases - on diagonal and not on diagonal

File: test/test-rotateElement.js (hunk 3 above)
Comment (PashaBeshchuk) [reply test/test-rotateElement.js]: I did not understand it:
"You can take zeroes element of the result of getCoordinates"

File: test/test-rotateElement.js (hunk 4 above)
Comment (zbstof) [reply test/test-rotateElement.js]: Add cases when element doesnt move after rotation. 3x3 and 5x5 cases

File: test/test-rotateElement.js [hunk 5]
```diff
@@ -0,0 +1,429 @@
...
+            -----
+            -----
+        `)
+        let stepElementTetromino = tetris.rotateElement(fulcrum[0], elementTetromino[0])
+		let expectedResult = tetris.getCoordinates(`
+            -----
+            -----
+            -----
+            -----
+            ----X
```
Comment (zbstof) [inline test/test-rotateElement.js]: I don't think this is correct

File: test/test-rotateElement.js [hunk 6]
```diff
@@ -0,0 +1,902 @@
...
+            -----
+            -----
+            -----
+            -----
+            ---X-
+        `)
+        assert.equal(expectedResult[0], stepElementTetromino)
+    })
+
+    it("the item returns to its original position", function(){
```
Comment (zbstof) [inline test/test-rotateElement.js]: these test names are meaningless

File: test/test-rotateElement.js [hunk 7]
```diff
@@ -0,0 +1,902 @@
...
+            -----
+            ----X
+            -----
+            -----
+            -----
+        `)
+        assert.equal(expectedResult[0], stepElementTetromino)
+    })
+
+describe("element rotation around the axis 5x5 not on diagonal", function () {
```
Comment (zbstof) [inline test/test-rotateElement.js]: I think this is exact copy of code starting from line 318 - please remove. All tests must be unique

File: test/test-rotateElement.js (hunk 6 above)
Comment (zbstof) [reply test/test-rotateElement.js]: Please rename all test cases so that:
1) they reflect exactly what is going on
2) they don't repeat across test suite (describe ...) - test names should be unique in entire file

File: test/test-rotateElement.js [hunk 8]
```diff
@@ -0,0 +1,902 @@
...
+            -----
+            ----X
+            -----
+            -----
+            -----
+        `)
+        assert.equal(expectedResult[0], stepElementTetromino)
+    })
+
+describe("element rotation around the axis 7x7 not on diagonal", function () {
```
Comment (zbstof) [inline test/test-rotateElement.js]: Your tests for 7x7 is exact copy of 5x5, but with 1 line of padding on the sides - not very useful - you're not covering new use-cases for your function

File: test/test-rotateElement.js (hunk 4 above)
Comment (zbstof) [reply test/test-rotateElement.js]: Please do this!

Comment (zbstof) [reply test/test-rotateElement.js]: > Add cases when element doesnt move after rotation. 3x3 and 5x5 cases

Please do this!

File: test/test-rotateElement.js (hunk 8 above)
Comment (PashaBeshchuk) [reply test/test-rotateElement.js]: There will be other coordinates

Comment (zbstof) [reply test/test-rotateElement.js]: yes, but I would rather see some more interesting tests

**Comment by zbstof** [comment]
-не квадратне поле
-точка оберту не по центру
-точка оберту співпадає з точкою що обертається
- при повороті точка виходить за межі поля (4 тести - на кожну сторону)

File: test/test-rotateElement.js [hunk 9]
```diff
@@ -0,0 +1,390 @@
...
+            --X--
+            -----
+            -----
+        `)
+        let stepElementTetromino = tetris.rotateElement(elementTetromino[0], pivot[0])
+        let expectedResult = tetris.getCoordinates(`
+            -----
+            -----
+            ----X
+        `)
```
Comment (zbstof) [inline test/test-rotateElement.js]: You forgot asserts - please add to all the tests

File: test/test-rotateElement.js [hunk 10]
```diff
@@ -0,0 +1,390 @@
...
+        let pivot = tetris.getCoordinates(`
+            --X--
+            -----
+            -----
+        `)
+        let stepElementTetromino = tetris.rotateElement(elementTetromino[0], pivot[0])
+        let expectedResult = [{ x: 4, y: -2 }]
```
Comment (zbstof) [inline test/test-rotateElement.js]: When you specify coordinates in non-graphical way - please use only object, no need to wrap it in the array

Comment (zbstof) [reply test/test-rotateElement.js]: Apply to all cases where you specify coordinates manually

File: test/test-rotateElement.js [hunk 11]
```diff
@@ -0,0 +1,390 @@
...
+        `)
+        let stepElementTetromino = tetris.rotateElement(elementTetromino[0], pivot[0])
+        let expectedResult = tetris.getCoordinates(`
+            -----
+            -----
+            ----X
+        `)
+    })
+
+    it("element move the top on the diagonal field 3x5 ", function () {
```
Comment (zbstof) [inline test/test-rotateElement.js]: you have unnecessary space at the end of names of most tests - please remove

File: test/test-rotateElement.js [hunk 12]
```diff
@@ -0,0 +1,390 @@
...
+        let pivot = tetris.getCoordinates(`
+            -----
+            -----
+            -X---
+            -----
+            -----
+            -----
+        `)
+        let stepElementTetromino = tetris.rotateElement(elementTetromino[0], pivot[0])
+        let expectedResult = [{ x: 1, y: -1 }]
```
Comment (zbstof) [inline test/test-rotateElement.js]: this is the same as elementTetromino, incorrect

Comment (zbstof) [reply test/test-rotateElement.js]: This whole test is just incorrect version of the next test

File: test/test-rotateElement.js [hunk 13]
```diff
@@ -0,0 +1,390 @@
...
+describe("Rotation where pivot coordinates coincides with element coordinates", function () {
+    it("element not move field 4x4 ", function () {
+        let elementTetromino = tetris.getCoordinates(`
+           ----
+           ----
+           -X--
+           ----
+        `)
+        let pivot = tetris.getCoordinates(`
+            ----
```
Comment (zbstof) [inline test/test-rotateElement.js]: please indent 1 space to the left - so it will be the same as elementTetromino

File: test/test-rotateElement.js [hunk 14]
```diff
@@ -0,0 +1,390 @@
...
+            -----
+        `)
+        let stepElementTetromino = tetris.rotateElement(elementTetromino[0], pivot[0])
+        let expectedResult = [{ x: -2, y: 2 }]
+    })
+})
+describe("Rotation where pivot coordinates coincides with element coordinates", function () {
+    it("element not move field 4x4 ", function () {
+        let elementTetromino = tetris.getCoordinates(`
+           ----
```
Comment (zbstof) [inline test/test-rotateElement.js]: please indent 1 space to the right - so it will be the same as pivot and expectedResult

File: tetris.js [hunk 15]
```diff
@@ -304,6 +304,8 @@ module.exports = {
 		}
 		return arrayOfCoordinates;
 
-	}
+	},
```
Comment (zbstof) [inline tetris.js]: I don't this this comma should be here

**Comment by zbstof** [comment]
Please edit the pull request name to:
1. Reflect what was done
2. Mention issue number for which this work was done

File: test/test-rotateElement.js (hunk 12 above)
Comment (PashaBeshchuk) [reply test/test-rotateElement.js]: does the error start with line 159 or 185?
I don't understand which test is wrong

Comment (zbstof) [reply test/test-rotateElement.js]: Remove test on line 340 - it is duplicate of the next test

Comment (PashaBeshchuk) [reply test/test-rotateElement.js]: error in line 354

File: test/test-rotateElement.js (hunk 14 above)
Comment (PashaBeshchuk) [reply test/test-rotateElement.js]: the pivot should space to the right?

Comment (zbstof) [reply test/test-rotateElement.js]: No, move each line of graphical representation of elementTetromino 1 space to the right

Comment (zbstof) [reply test/test-rotateElement.js]: Please resolve this as well

File: test/test-rotateElement.js [hunk 16]
```diff
@@ -0,0 +1,414 @@
...
+            -----
+            -----
+            -----
+        `)
+        assert.deepEqual(expectedResult[0], stepElementTetromino)
+    })
+})
+
+describe("Rotation where starts to the bottom, pivot has different x and y coordinates", function () {
+    it("element move the left field 6x6", function () {
```
Comment (zbstof) [inline test/test-rotateElement.js]: why did you change to 6x6?

File: test/test-rotateElement.js [hunk 17]
```diff
@@ -0,0 +1,414 @@
...
+            ------
+            --X---
+            ------
+            ------
+            ------
+        `)
+        let stepElementTetromino = tetris.rotateElement(elementTetromino[0], pivot[0])
+        let expectedResult = tetris.getCoordinates(`
+            ------
+            -----X
```
Comment (zbstof) [inline test/test-rotateElement.js]: this is not correct now

File: test/test-rotateElement.js [hunk 18]
```diff
@@ -0,0 +1,414 @@
...
+            ------
+            ------
+            ---X--
+        `)
+        assert.deepEqual(expectedResult, stepElementTetromino)
+    })
+})
+
+describe("Rotation where pivot coordinates coincides with element coordinates", function () {
+    it("element not move field 4x4", function () {
```
Comment (PashaBeshchuk) [inline test/test-rotateElement.js]: I do not understand
pivot === elementTetromino
and
expectedResult === pivot
too
What is wrong here?

Comment (zbstof) [reply test/test-rotateElement.js]: Pleaese google "indent" and make sure all 3 of these variables have the same indent
